File: java/eyes.playwright.java/src/main/java/com/applitools/eyes/playwright/universal/mapper/TFramesMapper.java
package com.applitools.eyes.playwright.universal.mapper;

import com.applitools.eyes.playwright.universal.Refer;
import com.applitools.eyes.playwright.universal.dto.Element;
import com.applitools.eyes.playwright.universal.dto.FrameLocator;
import com.applitools.eyes.playwright.universal.dto.Selector;
import com.applitools.eyes.universal.Reference;
import com.applitools.eyes.universal.dto.ContextReferenceDto;
import com.google.common.base.Strings;
import com.microsoft.playwright.ElementHandle;
import com.microsoft.playwright.Locator;

import java.util.List;
import java.util.stream.Collectors;

public class TFramesMapper {

    public static ContextReferenceDto toContextReferenceDto(FrameLocator frame, Refer refer, Reference root) {
        if (frame == null) {
            return null;
        }

        ContextReferenceDto contextReferenceDto = new ContextReferenceDto();

        String frameNameOrId = frame.getFrameNameOrId();
        if (!Strings.isNullOrEmpty(frameNameOrId)) {
            contextReferenceDto.setFrame(frameNameOrId);
        }

        Integer frameIndex = frame.getFrameIndex();
        if (frameIndex != null) {
            contextReferenceDto.setFrame(frameIndex);
        }

        Element frameElement = frame.getFrameElement();
        if (frameElement != null) {
            ElementHandle elementHandle = frameElement.getElementHandle();
            frameElement.setApplitoolsRefId(refer.ref(elementHandle, root));
            contextReferenceDto.setFrame(frameElement);
        }

        Selector frameSelector = frame.getFrameSelector();
        if (frameSelector != null) {
            if (frameSelector.getLocator() != null) {
                Locator locator = frameSelector.getLocator();
                frameSelector.setApplitoolsRefId(refer.ref(locator, root));
            }
            contextReferenceDto.setFrame(frameSelector);
        }

        Selector scrollSelector = frame.getScrollRootSelector();
        if (scrollSelector != null) {
            if (scrollSelector.getLocator() != null) {
                Locator locator = scrollSelector.getLocator();
                scrollSelector.setApplitoolsRefId(refer.ref(locator, root));
            }
            contextReferenceDto.setScrollRootElement(scrollSelector);
        }

        Element scrollElement = frame.getScrollRootElement();
        if (scrollElement != null) {
            ElementHandle elementHandle = scrollElement.getElementHandle();
            scrollElement.setApplitoolsRefId(refer.ref(elementHandle, root));
            contextReferenceDto.setScrollRootElement(scrollElement);
        }

        return contextReferenceDto;
    }

    public static List<ContextReferenceDto> toTFramesFromCheckSettings(List<FrameLocator> frameChain, Refer refer, Reference root) {
        if (frameChain == null || frameChain.isEmpty()) {
            return null;
        }

        return frameChain.stream()
                .map(frame -> toContextReferenceDto(frame, refer, root))
                .collect(Collectors.toList());
    }
}

File: dotnet/Eyes.Playwright.DotNet/Universal/Mapper/TFramesMapper.cs
using System.Collections.Generic;
using System.Linq;
using Applitools.Playwright.Universal.Dto;
using Applitools.Universal;
using Microsoft.Playwright;

namespace Applitools.Playwright.Universal.Mapper
{
    public static class TFramesMapper
    {
        public static ContextReferenceDto ToContextReferenceDto(FrameLocator frame, Refer refer, Reference root)
        {
            if (frame == null)
            {
                return null;
            }

            ContextReferenceDto contextReferenceDto = new ContextReferenceDto();

            string frameNameOrId = frame.FrameNameOrId;
            if (!string.IsNullOrEmpty(frameNameOrId))
            {
                contextReferenceDto.Frame = frameNameOrId;
            }

            int? frameIndex = frame.FrameIndex;
            if (frameIndex != null)
            {
                contextReferenceDto.Frame = frameIndex;
            }

            Element frameElement = frame.FrameElement;
            if (frameElement != null)
            {
                IElementHandle elementHandle = frameElement.ElementHandle;
                frameElement.ApplitoolsRefId = refer.Ref(elementHandle, root);
                contextReferenceDto.Frame = frameElement;
            }

            Selector frameSelector = frame.FrameSelector;
            if (frameSelector != null)
            {
                if (frameSelector.Locator != null)
                {
                    ILocator locator = frameSelector.Locator;
                    frameSelector.ApplitoolsRefId = refer.Ref(locator, root);
                }

                contextReferenceDto.Frame = frameSelector;
            }

            Selector scrollSelector = frame.ScrollRootSelector;
            if (scrollSelector != null)
            {
                if (scrollSelector.Locator != null)
                {
                    ILocator locator = scrollSelector.Locator;
                    scrollSelector.ApplitoolsRefId = refer.Ref(locator, root);
                }

                contextReferenceDto.ScrollRootElement = scrollSelector;
            }

            Element scrollElement = frame.ScrollRootElement;
            if (scrollElement != null)
            {
                IElementHandle elementHandle = scrollElement.ElementHandle;
                scrollElement.ApplitoolsRefId = refer.Ref(elementHandle, root);
                contextReferenceDto.ScrollRootElement = scrollElement;
            }

            return contextReferenceDto;
        }

        public static List<ContextReferenceDto> ToTFramesFromCheckSettings(IList<FrameLocator> frameChain, Refer refer,
            Reference root)
        {
            if (frameChain == null || frameChain.Count == 0)
            {
                return null;
            }

            return frameChain.Select(frame => ToContextReferenceDto(frame, refer, root)).ToList();
        }
    }
}